### PR TITLE
W-12637937:  AST Validations: Cannot inject required fields with minimal JSR-330 impl

### PR DESCRIPTION
**Pull request by mule-pachanady**
AST validations injection is intended to work with the minimal JSR-330 spec, which does not cover java.lang.Optional fields.

ArtifactAstValidationsTestCase uses feather fwk to inject the fields, which is a basic JSR-330 impl.

Before java 17, Feather worked fine with Optional because it could instrospct it and use its internal constructor. In java 17, it can no longer do that.
Replacing Feather with Guice for this test.

### CODE REVIEW

**Comment by mule-pachanady** [comment]
--validate

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 1]
```diff
@@ -66,9 +157,11 @@ public void before() throws Exception {
     extensionManager = new DefaultExtensionManager();
     muleContext.setExtensionManager(extensionManager);
     initialiseIfNeeded(extensionManager, muleContext);
+    injector = Guice.createInjector(new BasicModule());
```
Comment (fsgonz) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:160]: static import

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 2]
```diff
@@ -53,6 +84,66 @@
...
+        singletonMap(ENFORCE_EXPRESSION_VALIDATION, Boolean.valueOf(true));
+
+    @Override
+    protected void configure() {
+      bind(ArtifactAstValidatorBuilder.class).to(DefaultValidatorBuilder.class);
+      bind(ExtendedExpressionManager.class).to(DefaultExpressionManager.class);
+      bind(ValidationsProvider.class).to(CoreValidationsProvider.class);
+      bind(MuleContext.class).to(DefaultMuleContext.class);
+      bind(TransformersRegistry.class).to(DefaultTransformersRegistry.class);
+
```
Comment (fsgonz) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:104]: is this extra line necessary?

**Comment by mule-pachanady** [comment]
--validate

**Comment by mule-pachanady** [comment]
--validate

**Review by fsgonz — APPROVED** [review]

**Comment by mule-pachanady** [comment]
--validate

**Comment by mule-pachanady** [comment]
--validate

**Comment by mule-pachanady** [comment]
--validate

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 3]
```diff
@@ -53,6 +84,65 @@
...
+    @Override
+    protected void configure() {
+      bind(ArtifactAstValidatorBuilder.class).to(DefaultValidatorBuilder.class);
+      bind(ExtendedExpressionManager.class).to(DefaultExpressionManager.class);
+      bind(ValidationsProvider.class).to(CoreValidationsProvider.class);
+      bind(MuleContext.class).to(DefaultMuleContext.class);
+      bind(TransformersRegistry.class).to(DefaultTransformersRegistry.class);
+      bind(Object.class).to(String.class);
+      bind(Transformer.class).to(StringToBoolean.class);
```
Comment (elrodro83) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:105]: are these needed with the methods below?

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 4]
```diff
@@ -53,6 +84,65 @@
...
+    @Provides
+    @Singleton
+    public ExpressionLanguage provideExpressionLanguage() {
+      return new WeaveDefaultExpressionLanguageFactoryService(null).create();
+    }
+
+    @Provides
+    @Singleton
+    public FeatureFlaggingService provideFeatureFlaggingService() {
+      return new DefaultFeatureFlaggingService("abcd", featureBooleanMap);
```
Comment (elrodro83) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:117]: what do these parameters mean?

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 5]
```diff
@@ -53,6 +84,65 @@
 @Story(DSL_VALIDATION_STORY)
 public class ArtifactAstValidationsTestCase extends AbstractMuleContextTestCase {
 
+  private Injector injector;
+
+  @Rule
+  public SystemProperty systemProperty = new SystemProperty(ENFORCE_EXPRESSION_VALIDATION_PROPERTY, "true");
+
+  class BasicModule extends AbstractModule {
```
Comment (elrodro83) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:92]: the ides of this code apart from testing is that other teams use it as example reference, please add comments

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 6]
```diff
@@ -81,10 +173,8 @@ public void astValidationsWithBaseRegistryOutsideRuntime() throws ConfigurationE
   }
 
   protected List<ValidationResultItem> doValidate(ArtifactAst ast) throws ConfigurationException {
-    Feather feather = Feather.with(new BaseRegistryForValidationsModule());
-
     ArtifactAstValidator astValidator = validatorBuilder()
-        .withValidationEnricher(feather::injectFields)
+        .withValidationEnricher(p -> injector.injectMembers(p))
```
Comment (elrodro83) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:177]: use method reference

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 7]
```diff
@@ -53,6 +84,65 @@
 @Story(DSL_VALIDATION_STORY)
 public class ArtifactAstValidationsTestCase extends AbstractMuleContextTestCase {
 
+  private Injector injector;
+
+  @Rule
+  public SystemProperty systemProperty = new SystemProperty(ENFORCE_EXPRESSION_VALIDATION_PROPERTY, "true");
```
Comment (elrodro83) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:90]: why is this needed? isn't this the default?

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java (hunk 3 above)
Comment (mule-pachanady) [reply dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:105]: Sorry tried that, test wont run, got errors

```
Requested by:
1  : DefaultExpressionManager.setMuleContext(DefaultExpressionManager.java:379)
      \_ for 1st parameter muleContext
     at DefaultExpressionManager.setMuleContext(DefaultExpressionManager.java:379)
     at ArtifactAstValidationsTestCase$BasicModule.configure(ArtifactAstValidationsTestCase.java:100)

Learn more:
  https://github.com/google/guice/wiki/MISSING_IMPLEMENTATION

2) [Guice/MissingImplementation]: No implementation for TransformersRegistry was bound.

Requested by:
1  : DefaultExpressionManager.setTransformersRegistry(DefaultExpressionManager.java:384)
      \_ for 1st parameter transformersRegistry
     at DefaultExpressionManager.setTransformersRegistry(DefaultExpressionManager.java:384)
     at ArtifactAstValidationsTestCase$BasicModule.configure(ArtifactAstValidationsTestCase.java:100)
```

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java (hunk 4 above)
Comment (mule-pachanady) [reply dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:117]: We are just trying to verify all injections work and in valid DataWeave expression is validated, so we need to set ENFORCE_EXPRESSION_VALIDATION to true as it defaults to false, passing it via a map, artifact id here does not matter

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java (hunk 7 above)
Comment (mule-pachanady) [reply dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:90]: Actually this can be removed as we are injecting FeatureFlaggingService

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java (hunk 6 above)
Comment (mule-pachanady) [reply dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:177]: ok

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java (hunk 4 above)
Comment (mule-pachanady) [reply dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:117]: FeatureFlaggingService needs to be injected, but tryting just give it empty HashMap so that we dont set ENFORCE_EXPRESSION_VALIDATION to true, that fails the test as defaullts to false

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 8]
```diff
@@ -53,6 +84,67 @@
...
+    @Provides
+    @Singleton
+    Optional<FeatureFlaggingService> provideOptionalFeatureFlaggingService(FeatureFlaggingService featureFlaggingService) {
+      return Optional.of(featureFlaggingService);
+    }
+
+    @Provides
+    @Singleton
+    Collection<Transformer> provideCollectionTransformer() {
+      return new ArrayList<Transformer>();
```
Comment (elrodro83) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:138]: use `Collections.emptyList()`

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 9]
```diff
@@ -53,6 +84,67 @@
...
+    }
+
+    @Provides
+    @Singleton
+    List<TransformerResolver> provideListTransformerResolver() {
+      return new ArrayList<TransformerResolver>();
```
Comment (elrodro83) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:144]: use `Collections.emptyList()`

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 10]
```diff
@@ -53,6 +84,67 @@
...
+
+    @Provides
+    @Singleton
+    Collection<Transformer> provideCollectionTransformer() {
+      return new ArrayList<Transformer>();
+    }
+
+    @Provides
+    @Singleton
+    List<TransformerResolver> provideListTransformerResolver() {
```
Comment (elrodro83) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:143]: public

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 11]
```diff
@@ -53,6 +84,67 @@
...
+
+    @Provides
+    @Singleton
+    Optional<FeatureFlaggingService> provideOptionalFeatureFlaggingService(FeatureFlaggingService featureFlaggingService) {
+      return Optional.of(featureFlaggingService);
+    }
+
+    @Provides
+    @Singleton
+    Collection<Transformer> provideCollectionTransformer() {
```
Comment (elrodro83) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:137]: public

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 12]
```diff
@@ -53,6 +84,67 @@
...
+    @Provides
+    @Singleton
+    @Named("_compatibilityPluginInstalled")
+    Optional<Object> provideOptionalCompatibilityPluginInstalled(Object object) {
+      return Optional.of(object);
+    }
+
+    @Provides
+    @Singleton
+    Optional<FeatureFlaggingService> provideOptionalFeatureFlaggingService(FeatureFlaggingService featureFlaggingService) {
```
Comment (elrodro83) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:131]: public

File: dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java [hunk 13]
```diff
@@ -53,6 +84,67 @@
...
+    @Provides
+    @Singleton
+    public FeatureFlaggingService provideFeatureFlaggingService() {
+      return new DefaultFeatureFlaggingService("abcd", featureBooleanMap);
+    }
+
+    @Provides
+    @Singleton
+    @Named("_compatibilityPluginInstalled")
+    Optional<Object> provideOptionalCompatibilityPluginInstalled(Object object) {
```
Comment (elrodro83) [inline dsl/src/test/java/org/mule/test/config/ast/ArtifactAstValidationsTestCase.java:125]: public

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Review by elrodro83 — DISMISSED** [review]

**Review by elrodro83 — DISMISSED** [review]

**Comment by mule-pachanady** [comment]
--validate

**Comment by mule-pachanady** [comment]
--validate

**Comment by mule-pachanady** [comment]
--validate

**Review by elrodro83 — APPROVED** [review]

**Comment by mule-pachanady** [comment]
--validate
